akka-actor/src/main/scala/akka/pattern/AskSupport.scala

```diff
@@ -184,6 +300,42 @@ final class AskableActorRef(val actorRef: ActorRef) extends AnyVal {
 
 }
 
+/*
+ * Implementation class of the “ask” with explicit sender pattern enrichment of ActorRef
+ */
+final class ExplicitlyAskableActorRef(val actorRef: ActorRef) extends AnyVal {
+
+  def ask(message: ActorRef ⇒ Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
+    internalAsk(message, timeout, sender)
+
+  def ?(message: ActorRef ⇒ Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
+    internalAsk(message, timeout, sender)
+
+  /**
+   * INTERNAL API: for binary compatibility
+   */
+  private[pattern] def internalAsk(messageFactory: ActorRef ⇒ Any, timeout: Timeout, sender: ActorRef): Future[Any] = actorRef match {
+    case ref: InternalActorRef if ref.isTerminated ⇒
+      val message = messageFactory(ref.provider.deadLetters)
+      actorRef ! message
+      Future.failed[Any](new AskTimeoutException(s"""Recipient[$actorRef] had already been terminated. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
+    case ref: InternalActorRef ⇒
+      if (timeout.duration.length <= 0) {
+        val message = messageFactory(ref.provider.deadLetters)
+        Future.failed[Any](new IllegalArgumentException(s"""Timeout length must not be negative, question not sent to [$actorRef]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
+      } else {
+        val a = PromiseActorRef(ref.provider, timeout, targetName = actorRef, "unknown", sender)
+        val message = messageFactory(a)
+        a.messageClassName = message.getClass.getName
+        actorRef.tell(message, a)
+        a.result.future
+      }
+    case _ ⇒
+      val message = messageFactory(sender.asInstanceOf[InternalActorRef].provider.deadLetters)
```

Comment:
Using deadLetters like this is the right thing to do, it is consistent with the rest of Akka. However, this line will throw an NPE if no sender was given, so please check that case and bail out without invoking the function in this case.